pkg: add tweetnacl crypto library

kaspar030 opened the pull request:
I've played a bit with NaCl and wrote a simple example testing NaCl's public key crypto functions some years back.

Turned out that code runs unmodified with tweetnacl [1], which runs unmodified [2] on RIOT.

This PR adds tweetnacl as module and adds a simple public key creation/encryption/decryption example.

My example app uses malloc, which can probably be changed. Also, tweetnacl itself needs a lot of stack. I set (STACKSIZE_DEFAULT*4) for main_thread in the Makefile, but this will very probably fail on some platforms.

[1] http://tweetnacl.cr.yp.to/index.html
[2] I had to fix some signedness warnings. Patch is sent to DJB for upstream inclusion.

OlegHahm commented:
Have fun with writing documentation for the header!

*j/k* Nice idea! It's about time to add some stuff on the security site to RIOT.

Kijewski commented on sys/crypto/tweetnacl/LICENSE:
```diff
@@ -0,0 +1,3 @@
+From http://http://tweetnacl.cr.yp.to:
```
Typo

Kijewski commented on sys/crypto/tweetnacl/randombytes.c:
```diff
@@ -0,0 +1,27 @@
+void randombytes(uint8_t *target, uint64_t n)
+{
+    while (n>=4) {
+        uint32_t *pos = (uint32_t*) target;
```
I see alignment problems. Better memcpy.

Kijewski commented:
I did not look into the big file, yet. But this is really promising!

kaspar030 commented:
@Kijewski Regarding alignment, I changed it to use byte-wise copy. memcpy version was >60bytes larger.

OlegHahm commented:
@kaspar030, what's the state of this one?

kaspar030 commented:
apart from needing a rebase, last time a checked, the example worked fine at least on native an sam r21. That means the package is operational.

OlegHahm commented:
This one needs a reviewer!

OlegHahm commented:
@BytesGalore, I had the impression you were looking a bit into security lately. :)

OlegHahm commented:
We still need a reviewer for that. Any volunteers?

OlegHahm commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+#include "tweetnacl.h"
```
License is missing

miri64 replied:
https://github.com/RIOT-OS/RIOT/pull/4317/files#diff-0abdd814cde24db625913f773f8bf7e6R3 mentions it's public domain, but yeah a note in the C-file might be helpful, too.

OlegHahm commented:
Could you please uncrustify the big C file?

BytesGalore commented:
I will have a look this afternoon, (sry for the absence)

One thing I directly stumbled upon is a `malloc in the test but no free [1]
[1] https://github.com/RIOT-OS/RIOT/pull/4317/files#diff-e77e412b96be626119405b011444127aR29

OlegHahm commented:
Good catch: it seems that one could replace this malloc() could be easily replaced by a static array of variable size.

BytesGalore commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+#include "tweetnacl.h"
+#define FOR(i,n) for (i = 0;i < n;++i)
+#define sv static void
```
I would vote to replace such defines with their original keywords,
even if it grow larger than the original 100 tweets NaCl implementation

BytesGalore commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+  inv25519(x+32,x+32);
+  M(x+16,x+16,x+32);
+  pack25519(q,x+16);
+  return 0;
```
I guess such returns can be safely removed, when no other value is ever returned by the function

BytesGalore commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+
+int crypto_scalarmult_base(u8 *q,const u8 *n)
+{ 
+  return crypto_scalarmult(q,n,_9);
```
same here its always `0`, avoiding such returns would have slimmed the tweetnacl further

BytesGalore commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+  0x5b,0xe0,0xcd,0x19,0x13,0x7e,0x21,0x79
+} ;
+
+int crypto_hash(u8 *out,const u8 *m,u64 n)
```
I'm not against have an alternative, but we already have (documented) `hash` implementations so how about to make this one optional?

BytesGalore commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+
+  FOR(i,64) out[i] = h[i];
+
+  return 0;
```
again constant `0` return

BytesGalore commented on sys/crypto/tweetnacl/tweetnacl.c:
```diff
@@ -0,0 +1,810 @@
+  set25519(p[1],gf1);
+  set25519(p[2],gf1);
+  set25519(p[3],gf0);
+  for (i = 255;i >= 0;--i) {
```
why not use the `FOR` macro here? 
Or the other way around we could replace the macro with the original call.

BytesGalore replied:
ok strike this, the macro is forward iterating

BytesGalore commented on tests/tweetnacl/main.c:
```diff
@@ -0,0 +1,89 @@
+    const char message[] = "Test String";
+    struct identity id_alice;
+
+    printf("Creating keypair ALICE...\n");
```
`puts` please
and for the others too

BytesGalore commented on tests/tweetnacl/main.c:
```diff
@@ -0,0 +1,89 @@
+    printf("Done.\n");
+
+    const unsigned char *m = bytes_to_message(message,strlen(message)+1);
+    unsigned long long mlen = strlen(message)+crypto_box_ZEROBYTES+1;
```
I guess `int` would be sufficient

BytesGalore commented:
Honestly I don't know how to correctly review this PR, since its basically a package but directly integrated like a usual module.
The coding style is obviously non RIOT slimmed to the minimum and the provided documentation is a bit thin ;)

I mean we can focus on the provided test. 
In this respect I would propose to extend it to a unittest that covers all testable functions provided with the module.

kaspar030 commented:
well, yes. I created the module, and then copied a test program I once did for linux, and it worked unchanged for RIOT. So I PR'ed it.

I guess it doesn't make sense to review the upstream code...

-----Original Message-----
From: BytesGalore <notifications@github.com>
To: RIOT-OS/RIOT <RIOT@noreply.github.com>
Cc: Kaspar Schleiser <kaspar@schleiser.de>, Mention <mention@noreply.github.com>
Sent: Fr., 22 Juli 2016 22:38
Subject: Re: [RIOT-OS/RIOT] sys: crypto: add tweetnacl (#4317)

Honestly I don't know how to correctly review this PR, since its basically a package but directly integrated like a usual module.
The coding style is obviously non RIOT slimmed to the minimum and the provided documentation is a bit thin ;)

I mean we can focus on the provided test. 
In this respect I would propose to extend it to a unittest that covers all testable functions provided with the module.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.￼

BytesGalore commented:
>well, yes. I created the module, and then copied a test program I once did for linux, and it worked unchanged for RIOT. So I PR'ed it.

Which is totally cool, since it demonstrates the portability towards RIOT :D 

> I guess it doesn't make sense to review the upstream code...

Yeah I would get lost in the abbreviated functions/macros used there.
I tend to propose to use tweetnacl as package rather than a module, since if we integrate it directly to RIOT we should (I feel more like we MUST) at least ensure a certain level of quality regarding our coding conventions and documentation coverage.

BytesGalore commented:
I made an alternative PR #5687 providing `tweetnacl` as a package

kYc0o commented:
Feature freeze -> postponed.

OlegHahm commented:
I would propose the following:
* add the license to the C files
* uncrustify the code
* add some documentation stub, so that this module is at least showing in the documentation somehow

Fine with that, @kaspar030 and @BytesGalore?

BytesGalore commented:
@OlegHahm :+1:, lets give it a try :)

kYc0o commented:
@kaspar030 ping. Can you rebase for now? I'll try to make time and test it.

miri64 commented:
Also don't forget about the stuff @OlegHahm was suggesting ;-)

miri64 commented:
Postponed due to Feature Freeze

kYc0o commented:
I'm testing on OS X and I get a segfault while creating the identity, can you rebase and do the test?

kaspar030 commented:
> I'm testing on OS X and I get a segfault while creating the identity, can you rebase and do the test?

Yeah, there was a bug in randombytes(). That's fixed now.
I also factored the buffers out of main() so we can better measure the stacksize needed by tweetnacl (and removed the "struct id" cruft).

Could you try again?

kYc0o commented:
Actually I had found the problem and tested it again... (thinking you were maybe busy on other stuff). 

I'll test it asap.

kaspar030 commented:
Should I maybe move the tweetnacl-code into pkg/? I'm very opposed to fiddle with that code. It is half generated upstream by python code, and the API is (more or less) documented publicly. But as it is, it doesn't pass RIOT's coding conventions. Moving the original code to pkg/ would make this clearer...

kYc0o commented:
It's working now in native OS X. However, it crashes very bad in iotlab-m3. With  my previous patch (which sadly I didn't commit and drop it when you did the changes), but it was basically the same, with the random function adapted from #5687 (which was a package, btw)

kaspar030 commented:
> However, it crashes very bad in iotlab-m3.

Does it crash in randombytes? Works fine on sam21...

kYc0o commented:
Nope, it crashes while printing `ps()`

kYc0o commented:
sorry bad button :S

kaspar030 commented:
> Nope, it crashes while printing ps()

Ah, I just realized that the default stacksize for cortex-m is just 1k, so the hack in the makefile is not sufficient. will fix.

kYc0o commented:
Just did the test in a samr21-xpro and I experience similar problems (trash printed to STDOUT).

kaspar030 commented:
Please try with the latest commit, I adjusted the stacksize for main.

kYc0o commented:
Great! it works as expected. So would you like to make a pkg? I'm not against it and I agree that the code it's hard to change to RIOT conventions, but it seems that upstream changes are not very often (last change from 2014...).

BytesGalore commented:
@kYc0o we decided to closed #6072 which provided tweetnacl as package

BytesGalore commented:
sry, wrong pr number (fixed)

kYc0o commented:
Yes I know, but @kaspar030 seems to like the idea to move the code to pkg so that's why I asked...

miri64 commented:
I do not exactly understand the argument in #5687 for not including it as a pkg. The only con I see (to many downloads on CI) are basically true for any pkg.

kaspar030 commented:
The idea is to just move the code to pkg/, to make clear that its unmodified upstream code. I just tried, it looks weird to build that way.

kYc0o commented:
Well, if nobody complains about this "ugly" code for now, I volunteer to make a cleanup soon. Besides that I think this is ready for merging, so please squash and see what Murdock thinks.

kYc0o commented:
Maybe @OlegHahm [comment](https://github.com/RIOT-OS/RIOT/pull/4317#issuecomment-250320452) could be addressed and this PR can be merged soon?

kYc0o commented:
If @kaspar030 doesn't have time I can take over this PR, I started to make use of it so it would be nice to have it soon.

emmanuelsearch commented:
@kYc0o are you suggesting you could take over the PR ;)

kaspar030 commented:
- rebased
- moved code to pkg/

> add the license to the C files

The one from me has a license. for the original file, there's a LICENSE file.

> uncrustify the code

Don't want to modify the upstream code.

> add some documentation stub, so that this module is at least showing in the documentation somehow

did.

OlegHahm commented:
> Don't want to modify the upstream code.

Sure, that was my proposal in case we were not going for a package.

kYc0o commented:
Great! Sorry for the pressure :P

OlegHahm commented:
Hm, can we tag this package somehow so that we can make it a proper package once we "fixed" the packaging system, i.e. introduced some proper caching?

kaspar030 commented:
> Hm, can we tag this package somehow so that we can make it a proper package once we "fixed" the packaging system, i.e. introduced some proper caching?

Do you think that is necessary? If you think "third party code" instead of "package", it doesn't matter if we've actually imported the sources.

kYc0o commented:
Actually the code is not being downloaded from upstream (and they didn't touch the code since 2014, so I assume it's not maintained anymore) so I don't know if the goal is to put it on some other repo and pull it from there or pull it from upstream (hoping they're not going down)

kaspar030 commented:
> they didn't touch the code since 2014, so I assume it's not maintained
> anymore

With crypto libraries, this is considered a good sign. Code was flawless
for two years now. :)
And they'll never add features.

kYc0o commented:
>With crypto libraries, this is considered a good sign. Code was flawless
for two years now. :)
And they'll never add features.

Agreed ;-)

kYc0o commented:
Code still working properly on native and iotlab-m3. Please squash and let's see what Murdock has to say.

kYc0o commented:
Ah yes, just please change the commits and PR messages according to the new place on the tree.

OlegHahm commented:
@kaspar030 wrote:
> >  they didn't touch the code since 2014, so I assume it's not maintained
 anymore

> With crypto libraries, this is considered a good sign. Code was flawless
for two years now. :)

Nice metric for proving security... ;-)

OlegHahm commented:
> Do you think that is necessary? If you think "third party code" instead of "package", it doesn't matter if we've actually imported the sources.

Maybe it is not necessary, but if we would have proper caching for a package the effect would be almost zero, and just look much nicer.

kaspar030 commented:
> if we would have proper caching for a package the effect would be almost zero, and just look much nicer.

It would mean extra lines in the Makefiles, extra downloads for the cache-cold case, extra complexity, ...

OlegHahm commented:
It would also mean less code duplication, less ugly code in our beautiful repository, and a consistent style. And I don't see the extra complexity.

However, I do not insist and consider it a waste of time to discuss about such a bagatelle. Nevermind!

kYc0o commented on tests/tweetnacl/main.c:
```diff
@@ -0,0 +1,76 @@
+ */
+
+/**
+ * @ingroup     examples
```
Maybe a detail but the doc (if built) will be wrong...

kYc0o commented:
Some trailing white spaces and a use of malloc found by Murdock... I thought you were able to avoid malloc usage?

kYc0o commented:
@kaspar030 Can you fix the last Murdock complains? It's almost ready :)

kYc0o commented:
@kaspar030 ping!

emmanuelsearch commented:
@kaspar030 seems like a low hanging fruit?

kYc0o commented:
A fresh Murdock check shows now some cpp check issues... it seems that the headers need "CPP Compatibility" but we agreed on avoiding any modification to the sources, the same for the white spaces.

There's a way to "blacklist" this build for such checks?

For the missing malloc.h it's just a matter of erase that line, it's not needed.

OlegHahm commented:
@kYc0o, I just checked: the `dist/tools/cppcheck/check.sh` and `dist/tools/externc/check.sh` contain already the possibility to exclude a certain folder - you could add the same for `dist/tools/whitespacecheck/check.sh`.

kYc0o commented:
@OlegHahm So you propose to make a PR for such modifications before merge this one?

OlegHahm commented:
Another quickfix would be adding
`pkg/tweetnacl/*/* -diff`
to `.gitattributes`.

OlegHahm commented:
The following patch should eliminate most of Murdock's complaints:
https://gist.github.com/OlegHahm/b27a743a24213d9bf6e495cc1404cc8b

OlegHahm commented:
For the record: I don't like this workaround and would still prefer either a proper package or properly imported code.

kYc0o commented:
I prefer that one, there are several checks failing (cpp, doc, whitespace, etc...) so I think exclude `pkg/tweetnacl/src` and `pkg/tweetnacl/include` would be the best.

kYc0o commented:
Uhm... downloading the source code from upstream would solve Murdock complains?

OlegHahm commented:
Yes, external code is not checked by static tests.

kYc0o commented:
superseded by #6293
